File: Advanced/PageAndRDLX/CustomPdfExport/VB.NET/PdfRendering/Rendering/PdfInteractivityRenderer.vb
Imports System.Drawing
Imports GrapeCity.ActiveReports.Rendering.Export
Imports PdfSharp.Drawing
Imports PdfSharp.Pdf

Friend Class PdfInteractivityRenderer
    Implements IInteractivityRenderer

    Private ReadOnly _document As PdfDocument
    Private ReadOnly _outlines As IDictionary(Of String, PdfOutline) = New Dictionary(Of String, PdfOutline)()
    Private ReadOnly _graphics As IList(Of XGraphics)

    Public Sub New(document As PdfDocument, graphics As IList(Of XGraphics))
        _document = document
        _graphics = graphics
    End Sub

    Public Sub AddOutline(key As String, parentKey As String, targetPage As Integer, targetArea As RectangleF, name As String) Implements IInteractivityRenderer.AddOutline
        Dim outline As New PdfOutline(name, _document.Pages(targetPage - 1)) With {
            .PageDestinationType = PdfPageDestinationType.Xyz,
            .Left = targetArea.X / PdfConverter.TwipsPerPoint,
            .Top = targetArea.Y / PdfConverter.TwipsPerPoint
        }

        Dim parentOutline As PdfOutline = Nothing
        If _outlines.TryGetValue(parentKey, parentOutline) Then
            parentOutline.Outlines.Add(outline)
        Else
            _document.Outlines.Add(outline)
        End If
        _outlines(key) = outline
    End Sub

    Public Sub AddBookmark(key As String, sourcePage As Integer, sourceArea As RectangleF, targetPage As Integer, targetArea As RectangleF) Implements IInteractivityRenderer.AddBookmark
        Dim sourceRect As XRect = _graphics(sourcePage).Transformer.WorldToDefaultPage(XRect.FromLTRB(sourceArea.Left, sourceArea.Top, sourceArea.Right, sourceArea.Bottom))
        _document.Pages(sourcePage).AddDocumentLink(New PdfRectangle(sourceRect), targetPage)
    End Sub

    Public Sub UrlGoTo(link As String, sourcePage As Integer, sourceArea As RectangleF) Implements IInteractivityRenderer.UrlGoTo
        _document.Pages(sourcePage - 1).AddWebLink(New PdfRectangle(PdfConverter.Convert(sourceArea)), link)
    End Sub

    Public Sub DrillthroughGoTo(reportName As String, parameters As IDictionary(Of String, Object), sourcePage As Integer, sourceArea As RectangleF) Implements IInteractivityRenderer.DrillthroughGoTo
    End Sub

    Public Sub AddToggle(key As String, sourcePage As Integer, sourceArea As RectangleF) Implements IInteractivityRenderer.AddToggle
    End Sub

    Public Sub AddSorting(key As String, sourcePage As Integer, sourceArea As RectangleF) Implements IInteractivityRenderer.AddSorting
    End Sub
End Class

File: Advanced/PageAndRDLX/CustomPdfExport/C#/PdfRendering/Rendering/PdfInteractivityRenderer.cs
using System.Collections.Generic;
using System.Drawing;
using GrapeCity.ActiveReports.Rendering.Export;
using PdfSharp.Drawing;
using PdfSharp.Pdf;

namespace ActiveReports.Samples.Export.Rendering.PdfSharp
{
	internal class PdfInteractivityRenderer: IInteractivityRenderer
	{
		private readonly PdfDocument _document;
		private readonly IDictionary<string, PdfOutline> _outlines = new Dictionary<string, PdfOutline>();
		private readonly IList<XGraphics> _graphics;

		public PdfInteractivityRenderer(PdfDocument document, IList<XGraphics> graphics)
		{
			_document = document;
			_graphics = graphics;
		}
		
		public void AddOutline(string key, string parentKey, int targetPage, RectangleF targetArea, string name)
		{
			var outline = new PdfOutline(name, _document.Pages[targetPage - 1])
			{
				PageDestinationType = PdfPageDestinationType.Xyz,
				Left = targetArea.X / PdfConverter.TwipsPerPoint,
				Top = targetArea.Y / PdfConverter.TwipsPerPoint
			};
			
			if (_outlines.TryGetValue(parentKey, out var parentOutline))
				parentOutline.Outlines.Add(outline);
			else
				_document.Outlines.Add(outline);
			_outlines[key] = outline;
		}

		public void AddBookmark(string key, int sourcePage, RectangleF sourceArea, int targetPage, RectangleF targetArea)
		{
			var sourceRect = _graphics[sourcePage].Transformer.WorldToDefaultPage(XRect.FromLTRB(sourceArea.Left, sourceArea.Top, sourceArea.Right, sourceArea.Bottom));
			_document.Pages[sourcePage].AddDocumentLink(new PdfRectangle(sourceRect), targetPage);
		}

		public void UrlGoTo(string link, int sourcePage, RectangleF sourceArea)
		{
			_document.Pages[sourcePage - 1].AddWebLink(new PdfRectangle(PdfConverter.Convert(sourceArea)), link);
		}

		public void DrillthroughGoTo(string reportName, IDictionary<string, object> parameters, int sourcePage, RectangleF sourceArea)
		{ }

		public void AddToggle(string key, int sourcePage, RectangleF sourceArea)
		{ }

		public void AddSorting(string key, int sourcePage, RectangleF sourceArea)
		{ }
	}
}